src/repart/repart.c

```diff
@@ -7821,6 +7867,73 @@ static int parse_partition_types(const char *p, GptPartitionType **partitions, s
         return 0;
 }
 
+static int parse_join_signature(const char *p, Hashmap **verity_settings_map) {
+        _cleanup_(verity_settings_freep) VeritySettings *verity_settings = NULL;
+        _cleanup_free_ char *root_hash = NULL, *base = NULL, *signature = NULL;
+        _cleanup_free_ void *content = NULL;
+        size_t len;
+        int r;
+
+        assert(p);
+        assert(verity_settings_map);
+
+        r = extract_many_words(&p, ":", 0, &root_hash, &base, &signature);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse signature parameter '%s': %m", p);
+        if (r < 2 || r > 3)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Expected hash:sig");
+        if (r == 2) {
+                r = read_full_file(base, (char**) &content, &len);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to read root hash signature file '%s': %m", base);
+        } else {
+                if (!streq(base, "base64"))
+                        return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                               "Expected base64: prefix, got '%s'",
+                                               base);
+
+                r = unbase64mem(signature, &content, &len);
+                if (r < 0)
+                        return log_error_errno(r, "Failed to parse root hash signature '%s': %m", signature);
+        }
+        if (len == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL), "Empty verity signature specified.");
+        if (len > VERITY_SIG_SIZE)
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Verity signatures larger than %llu are not allowed.",
+                                       VERITY_SIG_SIZE);
+
+        verity_settings = new0(VeritySettings, 1);
+        if (!verity_settings)
+                return log_oom();
+
+        free_and_replace(verity_settings->root_hash_sig, content);
+        verity_settings->root_hash_sig_size = len;
+
+        r = unhexmem(root_hash, &content, &len);
+        if (r < 0)
+                return log_error_errno(r, "Failed to parse root hash '%s': %m", root_hash);
+        if (len < sizeof(sd_id128_t))
+                return log_error_errno(SYNTHETIC_ERRNO(EINVAL),
+                                       "Root hash must be at least 128-bit long: %s",
+                                       root_hash);
+
+        free_and_replace(verity_settings->root_hash, content);
+        verity_settings->root_hash_size = len;
+
+        r = hashmap_ensure_put(verity_settings_map,
+                               &verity_settings_hash_ops,
+                               root_hash,
```

Comment:
i am not a fan of this, as this means uppercase/lowercase issues. 

just turn this into a Set* instead of a Hashmap* and use a sparsely initialized VeritSettings structure as search key where only the two root hash fields are initialized

Reply:
ok